Add: allow rider to edit themselves.

teesloane opened the pull request:
Closes #249.

I noticed that in the routes file, the dispatch route for [editing a rider](https://github.com/bikebrigade/dispatch/blob/6d2c710e843325da2f9839c4da4623b3e8e5f6b5/lib/bike_brigade_web/router.ex#L153) has an action of `:edit`, whereas the route for [riders' action is ](https://github.com/bikebrigade/dispatch/blob/6d2c710e843325da2f9839c4da4623b3e8e5f6b5/lib/bike_brigade_web/router.ex#L103)`:edit_profile`
 
I assumed this was intentional, which is why I created a `save_rider_edit_impl`. It it's not, I can remove most of the code and make sure both pages use the same action.

mveytsman commented:
I think the intention was for `:edit_profile` to only let you save the stuff that riders can change, wheras `edit` let's you change anything.

I *think* the way this is writen, a rider could change the `is_dispatcher` flag on themselves if they crafted the request right, which is bad obviously.

Does that seem right? I think the pattern should be an admin changeset and a non admin profile changeset on Rider, and different `Rider.update...` calls for each path. Does that seem right?

teesloane commented:
> Does that seem right? I think the pattern should be an admin changeset and a non admin profile changeset on Rider, and different `Rider.update...` calls for each path. Does that seem right?

Yes, I think that makes sense. My default would have been to thinking that hiding the fields would be sufficient, but eliminating the possibility of that (whether it be bad actors or just bad luck) makes the most sense. I will look at the code again and let you know if I have any questions.

teesloane commented:
Updated this so that the changeset for editing of a ride profile differs based on the type of user making the edit.

mveytsman reviewed (approved):
I think this works.

To be totally honest I'm a little nervous that the changeset we differentiate is on the `RiderForm` not the `Rider` struct, but I don't have a better way yet -- if we ever get to the `User` modification we should look at this pattern again

teesloane commented:
@mveytsman I added some tests in this [commit](https://github.com/bikebrigade/dispatch/pull/250/commits/3753bd0dca35d808e62c8bdee0133932d4acc8b9).

Originally, I was going to test the actual trying to change the form, as a rider, and then log in as a dispatcher to make sure it hadn't changed, but it turned out that I couldn't even render the form with the unreachable fields in the first place. So, I added tests to assert that the correct error is raised and tested the inverse (that the admin-only fields only work for admins).

Lmk what you think.

mveytsman commented:
Ah interesting. It looks like LiveViewTest is not allowing us to submit the form with those fields filled in. So technically this test isn't testing what we want to test - e.g. the error comes from the test framework before it hits our code. So sorry to say this way of testing what we want to test is superflous.

I think it's fine to keep it, but technically it's kind of a test that will always pass if the form doesn't have the field, even if one could change it.

Does that make sense?

teesloane commented:
hmm, I have some questions about this, but they might be more core to how phoenix works with forms and the problem we are trying to avoid. Let's maybe leave this for now and pair on it when you have a moment?
